server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBuffer.java

```diff
@@ -161,11 +163,10 @@ public synchronized ShuffleDataResult getShuffleData(
       updateBufferSegmentsAndResultBlocks(
           lastBlockId, readBufferSize, bufferSegments, readBlocks, expectedTaskIds);
       if (!bufferSegments.isEmpty()) {
-        int length = calculateDataLength(bufferSegments);
-        byte[] data = new byte[length];
+        CompositeByteBuf byteBuf = Unpooled.compositeBuffer();
```

Comment:
Will `CompositeByteBuf` cause some extra copy because we may trigger resize action?
Will we use direct memory for `CompositeByteBuf`?

Reply:
CompositeByteBuf does not need to apply for memory. It depends on child ByteBuf. ByteBuf comes from netty server, so it depends on ByteBufAllocator in the end. I will fix this problem when init StreamServer.

Reply:
Actually PooledByteBufAllocator.DEFAULT use direct memory, we needn't fix any more.

Reply:
> Actually PooledByteBufAllocator.DEFAULT use direct memory, we needn't fix any more.
Do you consider `CompositeByteBuf#consolidateIfNeeded()`? In this method, if we exceed the maxComponentNum, we will trigger the logic of merging component.

Reply:
> CompositeByteBuf does not need to apply for memory. It depends on child ByteBuf. ByteBuf comes from netty server, so it depends on ByteBufAllocator in the end. I will fix this problem when init StreamServer.

Do you consider `CompositeByteBuf#consolidateIfNeeded()`? In this method, if we exceed the maxComponentNum, we will trigger the logic of merging component.

Reply:
You're right, we can apply for a CompositeByteBuf that uses off-heap memory. maxNumComponents defaults to 16. I don’t think we need to adjust it. When maxNumComponents is reached, zero-copy will be triggered. See CompositeByteBuf#consolidate0.

Reply:
CompositeByteBuf#consolidate0 call the method `transferTo`. The implement of transferTo is as below
<img width="545" alt="企业微信截图_85e7edc1-f8ec-486a-9a7f-49e6b4d86df7" src="https://github.com/apache/incubator-uniffle/assets/8159038/c3bcc5a5-3190-4e88-8a6a-25fb1e9c9322">
This seems a data copy.

Reply:
> CompositeByteBuf#consolidate0 call the method `transferTo`. The implement of transferTo is as below <img alt="企业微信截图_85e7edc1-f8ec-486a-9a7f-49e6b4d86df7" width="545" src="https://user-images.githubusercontent.com/8159038/238525070-c3bcc5a5-3190-4e88-8a6a-25fb1e9c9322.png"> This seems a data copy.

Should we boost maxNumComponents to avoid copy?

Reply:
I'm not sure, but there is a problem which we need to discuss. For me, we could add some comments, 16 is enough if we only use it to read memory data in my opinion.

Reply:
16 is not enough, because read buffer is 16M.

Reply:
> 16 is not enough, because read buffer is 16M.

We would better use a larger value to avoid data copy.

Reply:
Suppose the size of a block is small enough to be 16k, readBuffer=16M, we can make maxNumComponents=1024 to avoid data copy.

Reply:
OK.